analysis-engine/src/test/java/org/sonarsource/sonarlint/core/analysis/mediumtests/AnalysisEngineMediumTests.java

```diff
@@ -82,12 +98,116 @@ void should_analyze_a_single_file_outside_of_any_module(@TempDir Path baseDir) t
+      .addInputFiles(inputFile)
+      .addActiveRules(trailingCommentRule())
+      .setBaseDir(baseDir)
+      .build();
+    List<Issue> issues = new ArrayList<>();
+    analysisEngine.post(new RegisterModuleCommand(new ClientModuleInfo("moduleKey", aModuleFileSystem())), progressMonitor).get();
+    analysisEngine.post(new AnalyzeCommand("moduleKey", analysisConfig, issues::add, null), progressMonitor).get();
+    assertThat(issues)
+      .extracting("ruleKey", "message", "inputFile", "flows", "quickFixes", "textRange.startLine", "textRange.startLineOffset", "textRange.endLine", "textRange.endLineOffset")
+      .containsOnly(tuple("python:S139", "Move this trailing comment on the previous empty line.", inputFile, List.of(), List.of(), 2, 9, 2, 27));
+  }
+
+  @Test
+  void should_fail_the_future_if_the_command_execution_fails() {
+    var futureResult = analysisEngine.post((moduleRegistry, progress) -> {
+      throw new RuntimeException("Kaboom");
+    }, progressMonitor);
+
+    Awaitility.await().until(futureResult::isCompletedExceptionally);
+    futureResult.exceptionally(e -> assertThat(e)
+      .isInstanceOf(RuntimeException.class)
+      .hasMessage("Kaboom"));
+  }
+
+  @Test
+  void should_execute_pending_commands_when_gracefully_finishing(@TempDir Path baseDir) throws IOException {
+    var futureRegister = analysisEngine.post(new RegisterModuleCommand(new ClientModuleInfo("moduleKey", aModuleFileSystem())), progressMonitor);
+    var futureNotify = analysisEngine.post(new NotifyModuleEventCommand("moduleKey", ClientModuleFileEvent.of(preparePythonInputFile(baseDir, ""), ModuleFileEvent.Type.CREATED)),
+      progressMonitor);
+    var futureUnregister = analysisEngine.post(new UnregisterModuleCommand("moduleKey"), progressMonitor);
+    var futureLongCommand = analysisEngine.post((moduleRegistry, progressMonitor) -> {
+      try {
+        Thread.sleep(1000);
+      } catch (InterruptedException e) {
+        e.printStackTrace();
+      }
+      return "SUCCESS";
+    }, progressMonitor);
+    analysisEngine.finishGracefully();
+    engineStopped = true;
+
+    Awaitility.await().until(futureLongCommand::isDone);
+    assertThat(futureLongCommand).isCompletedWithValue("SUCCESS");
+    assertThat(futureRegister).isCompleted();
+    assertThat(futureNotify).isCompleted();
+    assertThat(futureUnregister).isCompleted();
+  }
+
+  @Test
+  void should_cancel_progress_monitor_of_executing_command_when_stopping() throws InterruptedException {
+    var futureLongCommand = analysisEngine.post((moduleRegistry, progressMonitor) -> {
+      int attempts = 0;
+      while (attempts < 10 && !progressMonitor.isCanceled()) {
+        try {
+          Thread.sleep(500);
+          attempts++;
+        } catch (InterruptedException e) {
+          // ignore
+        }
+      }
+      if (!progressMonitor.isCanceled()) {
+        fail("Progress monitor has not been canceled properly");
+      }
+      return "SUCCESS";
```

Comment:
I would change `"SUCCESS"` to `"CANCELLED"` for clarity of what the test is doing